rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,994 @@
+              arr(pos0) = numItem.toByte
+              // Fill second byte - NodePtr identifier (most significant bit = 1) and prefixSize (lower 7 bits = size)
+              val posSecondByte   = pos0 + 1
+              val prefixSize: Int = prefix.size.toInt
+              arr(posSecondByte) = encodeNodePtrSecondByte(prefixSize)
+              // Fill prefix
+              val posPrefixStart = posSecondByte + 1
+              for (i <- 0 until prefixSize) arr(posPrefixStart + i) = prefix(i.toLong)
+              // Fill ptr
+              val posPtrStart = posPrefixStart + prefixSize
+              for (i <- 0 until defSize) arr(posPtrStart + i) = ptr(i.toLong)
+              putItemIntoArray(numItem + 1, posPtrStart + defSize) // Loop to the next item.
+          }
+        }
+      ByteVector(putItemIntoArray(0, 0))
+    }
+
+    /** Deserialization [[ByteVector]] to [[Node]]
+      */
+    def decode(bv: ByteVector): Node = {
+      val arr     = bv.toArray
+      val maxSize = arr.length
+
+      // If first bit 0 - return true, otherwise false.
+      def isLeaf(secondByte: Byte) = (secondByte & 0x80) == 0x00
+
+      @tailrec
+      // Each loop decodes one non-empty item.
+      def decodeItem(pos0: Int, node: Node): Node =
+        if (pos0 == maxSize) node // End of deserialization
+        else {
+          val (idx0Next, nodeNext) = try {
+            val numItem: Int = byteToInt(arr(pos0)) // Take first byte - it's item's number
+            assert(
+              node(numItem) == EmptyItem,
+              "Error during deserialization: wrong number of item."
+            )
+            val pos1       = pos0 + 1
+            val secondByte = arr(pos1) // Take second byte
+
+            // Decoding prefix
+            val prefixSize: Int = secondByte & 0x7F // Lower 7 bits - it's size of prefix (0..127).
+            val prefix          = new Array[Byte](prefixSize)
+            val posPrefixStart  = pos1 + 1
+            for (i <- 0 until prefixSize) prefix(i) = arr(posPrefixStart + i) // Take prefix
+
+            // Decoding leaf or nodePtr data
+            val valOrPtr         = new Array[Byte](defSize)
+            val posValOrPtrStart = posPrefixStart + prefixSize
+            for (i <- 0 until defSize)
+              valOrPtr(i) = arr(posValOrPtrStart + i) // Take next 32 bytes - it's data
+
+            val pos0Next = posValOrPtrStart + defSize // Calculating start position for next loop
+
+            // Decoding type of non-empty item
+            val item = if (isLeaf(secondByte)) {
+              Leaf(ByteVector(prefix), ByteVector(valOrPtr))
+            } else {
+              NodePtr(ByteVector(prefix), ByteVector(valOrPtr))
+            }
+
+            (pos0Next, node.updated(numItem, item))
+          } catch {
+            case _: Exception =>
+              assert(assertion = false, "Error during deserialization: invalid data format")
+              (maxSize, emptyNode)
```

Comment:
The problem here is that original error is swallowed and assert is misused to throw errors. If you want to rethrow error you need to wrap the whole expression in F type. Or just let original error be thrown.

I think we can just remove try/catch block and add assertions where you think is needed to detect invalid data format.

Reply:
It seems that `Codec.decode` is already called in functions with F context so it make sense to convert this function also.

Reply:
I try to fix it but I have slowdiwn 10% for read operation with this code: 
```scala
...
      def invalidDataException: F[Unit] =
        new RuntimeException(s"Error during deserialization: invalid data format").raiseError
      for {
        r <- Sync[F].delay(decodeItem(0, emptyNode)).attempt
        _ <- invalidDataException.whenA(r.isLeft)
      } yield r.right.get
    }
  }
```
So, I am deleting `try catch` and don't intercepting this error.